packages/block-library/src/image/index.php

```diff
@@ -59,48 +59,54 @@ function render_block_core_image( $attributes, $content ) {
 		}
 		$content = $processor->get_updated_html();
 
+		$w = new WP_HTML_Tag_Processor( $content );
+		$w->next_tag(
+			array(
+				'tag_name' => 'figure',
+			)
+		);
+		$w->add_class( 'wp-lightbox-container' );
+		$w->set_attribute( 'data-wp-island', '' );
```

Comment:
this is going to produce `data-wp-island=""` but if all we need is the attribute as a boolean we can pass `set_attribute( 'data-wp-island', true )` and it will generate `data-wp-island`

in JS and in the browser both should be equivalent

Reply:
Funny thing, the browser works the other way around:

- `element.setAttribute("data-wp-island", "")` -> `data-wp-island`
- `element.setAttribute("data-wp-island", "true")` -> `data-wp-island="true"`

Reply:
But FWIW, I think `true` makes more sense. That's the way Preact/React work, so that's what we did for the Interactivity API. I'm happy that the HTML Tag Processor does the same thing 😄

We have an exception, though: `data-` and `aria-` attributes. Those should be enumerated. If you pass `true` or `false`, instead of adding/removing the element, it writes it as a string. It makes sense when working with things like `aria-expanded`, where you need the attribute to be either `aria-expanded="true"` and `aria-expanded="false"` instead of appear/disappear. Forcing people to remember to make a selector just to turn `true`/`false` into `"true"`/`"false"` could cause accessibility problems.

Reply:
@luisherranz it's of note that `"true"` is a different value than `""`, which is why the Tag Processor doesn't output `"true"`. it's really complicated.

for `aria-expanded` this is fine because the spec wants a string value of `"true"`, but if something is looking for a boolean HTML attribute and reads `"false"` it will consider it true 🙃. an example of this is setting `disabled="false"`, which disables the element it's on.

`true` is added in the Tag Processor so that it's possible to differentiate an attribute with an empty string value and an attribute with no value. similarly if you pass `false` as the value it will remove the attribute entirely, which is the only way to do things like set `disabled` to `false`.

just wanted to share; the code works as-is.

Reply:
sidenote: part of why it's confusing is that all HTML attribute values are strings, so the only way to communicate a boolean value is by the presence or absence of any value.

in addition, if the attribute _has_ a value, then it must be exactly the empty string `""` or exactly a case-insensitive match for the attribute's canonical name, like `"disabled"`

> Forcing people to remember to make a selector just to turn true/false into "true"/"false" could cause accessibility problems.

I agree, but that's kind of my point as well. this is water under the bridge because different HTML attributes demand that we know the difference and use the right one; that's not caused by Preact or the Tag Processor - those are just trying their best to conform to what is required so that they don't unintentionally flip a boolean value.

Reply:
> if all we need is the attribute as a boolean we can pass `set_attribute( 'data-wp-island', true )` and it will generate `data-wp-island` [...] in JS and **in the browser both should be equivalent**

What I was trying to say is that that's not how the browser works. In the browser, `elem.setAttribute("data-wp-island", true)` produces `data-wp-island="true"`, not `data-wp-island`. To produce `data-wp-island` you need to use `elem.setAttribute("data-wp-island", "")`.

Reply:
we're _almost_ saying the same thing I think 😄 

_but_! be careful with the JS in the browser and using `setAttribute( …, false )` because it's not doing what the Tag Processor is doing. its behavior follows type casting, evident when setting the value to `null`

if you call `element.setAttribute( 'disabled', false )` you will accidentally disable the element you meant to enable. in those cases the only way to disable the element is to call `element.removeAttrbiute( 'disabled' )`

so this is the inherent danger that I was referring to, because JS and the DOM API is the source of the problem.

Reply:
> because JS and the DOM API is the source of the problem

That's the reason [I was happy](https://github.com/WordPress/gutenberg/pull/51089#discussion_r1211939558) about the Tag Processor API 😄